### PR TITLE
diff: Fix broken `Local()` logic

### LINKED ISSUE #61145: kubectl diff feature always returns empty
/kind bug
/sig cli
/area kubectl

Hello.  This is my first interaction with the Kubernetes community, so please forgive me if i have left out any crucial details, or if, perhaps, this is PEBKAC after all!

I am keen to have a working `kubectl diff` feature, as described in https://github.com/kubernetes/kubernetes/issues/6284 and https://github.com/kubernetes/features/issues/491.  I am running K8s v1.9.3 and realise that this feature is in alpha, but i thought i would ask anyway.  We're not on v1.9.4 yet, but the changelog doesn't mention the `diff` feature.

**What you expected to happen**:

I have some services deployed to Kubernetes with YAML files, that has been working reliably.  We always use `kubectl apply -f ...` to deploy things.  When a service is already deployed, and i make a minor change in the config file, for example the service port, i expect to see that in the diff.  Instead, `kubectl` gives no output and exits with a `0`.

**How to reproduce it (as minimally and precisely as possible)**:

Consider the file `kube2iam.yaml`:
```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: kube2iam
  namespace: kube-system
  labels:
    app: kube2iam
spec:
  template:
    metadata:
      labels:
        name: kube2iam
    spec:
      hostNetwork: true
      tolerations:
      - key: "role"
        operator: "Equal"
        value: "controller"
      containers:
        - image: jtblin/kube2iam:latest
          name: kube2iam
          args:
            - "--auto-discover-default-role"
            - "--verbose"
          ports:
            - containerPort: 8181
              hostPort: 8181
              name: http
```
I make sure that it is correctly running on the cluster:
```sh
kubectl apply -f ./kube2iam.yaml
daemonset "kube2iam" unchanged
```
Great, looking good.  Now, i change the `spec.template.spec.containers[0].image` value to use Docker image tag `master` and save.

Now, i run:
```sh
kubectl alpha diff -f ./kube2iam.yaml
```

The stderr/stdout output is empty, and the exit code is `0`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version:

Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-09T21:50:44Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3+coreos.0", GitCommit:"f588569ed1bd4a6c986205dd0d7b04da4ab1a3b6", GitTreeState:"clean", BuildDate:"2018-02-10T01:42:55Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

### PR DESCRIPTION
**Pull request by apelisse**
Local and Live functions where doing and returning the same thing,
giving empty results by default. Fix the local function by copying the
objects before fetching the live version.

**What this PR does / why we need it**: Diff prints empty output by default. Fixes it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubernetes/issues/61145

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
Test case so this doesn't revert in future?

**Comment by apelisse** [comment]
I probably could but:
1. I'm not sure how to mock/fake/test the interaction with the resourcebuilder,
2. I don't want to invest too much time working on this, since it will have to change with server-side apply.

**Comment by toothbrush** [comment]
Is there any chance this will make it into a minor 1.9 or 1.10 release?  I'm really keen on this feature, and although i realise the implementation will have to change soon anyway, it'd be really nice to have in the interim.

**Comment by apelisse** [comment]
@toothbrush Actually, it won't have to change that much, so we should find a path forward at this point. I don't really have time to figure out how to test this, but I'd really like to get this merged too.

@smarterclayton Any suggestions on how we can test this? :-)

**Comment by smarterclayton** [comment]
Why is mocking required?  I was expecting a test-cmd scenario that prevents this regressing when we have server side apply

**Comment by apelisse** [comment]
> Why is mocking required? I was expecting a test-cmd scenario that prevents this regressing when we have server side apply

Sounds good. Thanks!

**Comment by toothbrush** [comment]
I wish i could contribute more to this discussion than simply "how are we doing?" but i'm afraid that's all i've got!  Maybe i should try my hand at constructing the test command?  I would need pointers from the experts, though.

**Comment by apelisse** [comment]
I'll work on this tomorrow

**Comment by apelisse** [comment]
@smarterclayton I've added a test. It's not the best test in the world, but it would have caught this bug.

File: pkg/kubectl/cmd/diff.go [hunk 1]
```diff
@@ -417,25 +453,29 @@ func RunDiff(f cmdutil.Factory, diff *DiffProgram, options *DiffOptions, from, t
...
 		}
 
-		if err := info.Get(); err != nil {
-			if !errors.IsNotFound(err) {
-				return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving current configuration of:\n%s\nfrom server for:", info.String()), info.Source, err)
-			}
-			info.Object = nil
+		remote, err := dl.Download(info)
+		if err != nil {
+			glog.Warningf("failed to download remote object: %v", err)
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/diff.go]: Nit but human warnings aren't via glog, we use `fmt.Fprintf(stdErr, "warning: ...")`

glog is terrible for humans.

File: pkg/kubectl/cmd/diff.go [hunk 2]
```diff
@@ -390,6 +382,50 @@ func (d *Differ) TearDown() {
...
+	if err != nil {
+		return nil, err
+	}
+	d.ns, _, _ = f.DefaultNamespace()
+
+	return &d, nil
+}
+
+func (d *Downloader) Download(info *resource.Info) (*unstructured.Unstructured, error) {
+	gvk := info.Object.GetObjectKind().GroupVersionKind()
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/diff.go]: @deads2k is any of this impacted by your mapper changes?

File: pkg/kubectl/cmd/diff.go [hunk 3]
```diff
@@ -263,6 +266,7 @@ type Object interface {
 // InfoObject is an implementation of the Object interface. It gets all
 // the information from the Info object.
 type InfoObject struct {
+	Remote  *unstructured.Unstructured
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/diff.go]: adding remote here as unstructured explicitly seems inconsistent.  I would have expected

```
Remote runtime.Unstructured
```

at a minimum.

File: pkg/kubectl/cmd/diff.go (hunk 2 above)
Comment (deads2k) [reply pkg/kubectl/cmd/diff.go]: > @deads2k is any of this impacted by your mapper changes?

This should be ok.  The way this is used you always end up with a GVK.

**Pull request merged**